substrate/frame/identity/src/types.rs

```diff
@@ -228,143 +230,31 @@ impl<Balance: Encode + Decode + MaxEncodedLen + Copy + Clone + Debug + Eq + Part
-			.type_params(vec![TypeParameter::new("T", Some(meta_type::<IdentityField>()))])
-			.composite(Fields::unnamed().field(|f| f.ty::<u64>().type_name("IdentityField")))
-	}
-}
-
 /// Information concerning the identity of the controller of an account.
-///
-/// NOTE: This should be stored at the end of the storage item to facilitate the addition of extra
-/// fields in a backwards compatible way through a specialized `Decode` impl.
-#[derive(
-	CloneNoBound, Encode, Decode, Eq, MaxEncodedLen, PartialEqNoBound, RuntimeDebugNoBound, TypeInfo,
-)]
-#[codec(mel_bound())]
-#[cfg_attr(test, derive(frame_support::DefaultNoBound))]
-#[scale_info(skip_type_params(FieldLimit))]
-pub struct IdentityInfo<FieldLimit: Get<u32>> {
-	/// Additional fields of the identity that are not catered for with the struct's explicit
-	/// fields.
-	pub additional: BoundedVec<(Data, Data), FieldLimit>,
-
-	/// A reasonable display name for the controller of the account. This should be whatever it is
-	/// that it is typically known as and should not be confusable with other entities, given
-	/// reasonable context.
-	///
-	/// Stored as UTF-8.
-	pub display: Data,
-
-	/// The full legal name in the local jurisdiction of the entity. This might be a bit
-	/// long-winded.
-	///
-	/// Stored as UTF-8.
-	pub legal: Data,
-
-	/// A representative website held by the controller of the account.
-	///
-	/// NOTE: `https://` is automatically prepended.
-	///
-	/// Stored as UTF-8.
-	pub web: Data,
-
-	/// The Riot/Matrix handle held by the controller of the account.
-	///
-	/// Stored as UTF-8.
-	pub riot: Data,
-
-	/// The email address of the controller of the account.
-	///
-	/// Stored as UTF-8.
-	pub email: Data,
-
-	/// The PGP/GPG public key of the controller of the account.
-	pub pgp_fingerprint: Option<[u8; 20]>,
-
-	/// A graphic image representing the controller of the account. Should be a company,
-	/// organization or project logo or a headshot in the case of a human.
-	pub image: Data,
-
-	/// The Twitter identity. The leading `@` character may be elided.
-	pub twitter: Data,
-}
-
-impl<FieldLimit: Get<u32>> IdentityInfo<FieldLimit> {
-	pub(crate) fn fields(&self) -> IdentityFields {
-		let mut res = <BitFlags<IdentityField>>::empty();
-		if !self.display.is_none() {
-			res.insert(IdentityField::Display);
-		}
-		if !self.legal.is_none() {
-			res.insert(IdentityField::Legal);
-		}
-		if !self.web.is_none() {
-			res.insert(IdentityField::Web);
-		}
-		if !self.riot.is_none() {
-			res.insert(IdentityField::Riot);
-		}
-		if !self.email.is_none() {
-			res.insert(IdentityField::Email);
-		}
-		if self.pgp_fingerprint.is_some() {
-			res.insert(IdentityField::PgpFingerprint);
-		}
-		if !self.image.is_none() {
-			res.insert(IdentityField::Image);
-		}
-		if !self.twitter.is_none() {
-			res.insert(IdentityField::Twitter);
-		}
-		IdentityFields(res)
+pub trait IdentityInformationProvider:
+	Encode + Decode + MaxEncodedLen + Clone + Debug + Eq + PartialEq + TypeInfo
```

Comment:
Why is this in storage and not a static type? Does it really need the `self` pointer?

Reply:
I think you mean `fn has_identity(&self, fields: u64) -> bool;` which does need `self` because it's an instance query, i.e. there are multiple instances of a `struct` implementing this and some might have a subset of fields enabled, some might not. The function is not asking if the type has a field, it's asking if the instance has that field set in the bit flags. See usage [here](https://github.com/paritytech/polkadot-sdk/pull/1661/files#diff-4a95a3948151a729b1fa4e91da9c72cdc6cd2d1611842fca6f177efeed34fd1aR941) and [here](https://github.com/paritytech/polkadot-sdk/blob/c86b633695299ed27053940d5ea5c5a2392964b3/substrate/frame/alliance/src/lib.rs#L140).